File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/flutterobject/NativeObjectStorage.java
package com.valentingrigorean.arcgis_maps_flutter.flutterobject;

import androidx.annotation.NonNull;

import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class NativeObjectStorage {
    private static final NativeObjectStorage instance = new NativeObjectStorage();
    private HashMap<String, NativeObject> nativeObjects = new HashMap<>();

    private NativeObjectStorage() {

    }

    public static NativeObjectStorage getInstance() {
        return instance;
    }

    public void addNativeObject(@NonNull NativeObject object) {
        nativeObjects.putIfAbsent(object.getObjectId(), object);
    }

    public void removeNativeObject(@NonNull String objectId) {
        NativeObject object = nativeObjects.remove(objectId);
        if (object != null) {
            object.dispose();
        }
    }

    public NativeObject getNativeObject(@NonNull String objectId) {
        return nativeObjects.get(objectId);
    }

    public void clearAll() {
        nativeObjects.forEach((key, value) -> value.dispose());
        nativeObjects.clear();
    }


    public static <T> T getNativeObjectOrConvert(Object object, Function<Object, T> mappingFunction) {
        final Map<?, ?> data = Convert.toMap(object);
        final String objectId = (String) data.get("nativeObjectId");
        if (objectId != null) {
            final NativeObject nativeObject = NativeObjectStorage.getInstance().getNativeObject(objectId);
            if (nativeObject != null) {
                final BaseNativeObject<T> baseNativeObject = (BaseNativeObject<T>) nativeObject;
                return baseNativeObject.getNativeObject();
            }
        }
        return mappingFunction.apply(object);
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/flutterobject/NativeObjectStorage.kt
package com.valentingrigorean.arcgis_maps_flutter.flutterobject
import java.util.function.Function

class NativeObjectStorage private constructor() {
    private val nativeObjects = HashMap<String, NativeObject>()
    fun addNativeObject(nativeObject: NativeObject) {
        nativeObjects.putIfAbsent(nativeObject.objectId, nativeObject)
    }

    fun removeNativeObject(objectId: String) {
        val `object` = nativeObjects.remove(objectId)
        `object`?.dispose()
    }

    fun<T : NativeObject> getNativeObject(objectId: String): T {
        return nativeObjects[objectId] as T
    }

    fun clearAll() {
        nativeObjects.forEach { (_: String, value: NativeObject) -> value.dispose() }
        nativeObjects.clear()
    }

    companion object {
        val instance = NativeObjectStorage()
        fun <T> getNativeObjectOrConvert(obj: Any, mappingFunction: Function<Any?, T>): T? {
            val data = obj as Map<*, *>
            val objectId = data["nativeObjectId"] as String?
            if (objectId != null) {
                val nativeObject = instance.getNativeObject<NativeObject>(objectId)
                if (nativeObject != null) {
                    val baseNativeObject = nativeObject as BaseNativeObject<T>
                    return baseNativeObject.nativeObject
                }
            }
            return mappingFunction.apply(obj)
        }
    }
}